primitives/src/merkle_tree/field_based_mht/smt/big_lazy_merkle_tree.rs

```diff
@@ -0,0 +1,987 @@
+                }
+            }
+            nodes_to_process_in_parallel.push(higher_level_nodes);
+            height += 1;
+        }
+
+        // Inserts leaves in the db and marks as present/not present.
+        for i in 0..vec_leaf_op.len() {
+            let x = vec_leaf_op[i];
+            let action = x.action;
+            let coord = x.coord;
+            let hash = x.hash;
+            let idx = coord.idx;
+
+            if action == Remove {
+                self.remove_from_db(idx);
+                self.state.present_node.remove(&coord);
+            } else {
+                self.insert_to_db(idx, hash.unwrap());
+                self.state.present_node.insert(coord);
+            }
+        }
+
+        // Compute hashes in parallel - first level
+        let mut input_vec = Vec::new();
+        let mut both_children_present = Vec::new();
+
+        for j in 0..nodes_to_process_in_parallel[0].len() {
+            let coord = nodes_to_process_in_parallel[0][j];
+
+            let idx = coord.idx;
+            let left_child_idx = idx * T::MERKLE_ARITY;
+            let right_child_idx= left_child_idx + 1;
+            let left_hash;
+            let right_hash;
+            let left_child_present;
+            let right_child_present;
+            let left_leaf = self.get_from_db(left_child_idx);
+            if let Some(i) = left_leaf {
+                left_hash = i;
+                left_child_present = true;
+            } else {
+                left_hash = T::EMPTY_HASH_CST[0];
+                left_child_present = false;
+            }
+
+            let right_leaf = self.get_from_db(right_child_idx);
+            if let Some(i) = right_leaf {
+                right_hash = i;
+                right_child_present = true;
+            } else {
+                right_hash = T::EMPTY_HASH_CST[0];
+                right_child_present = false;
+            }
+            input_vec.push(left_hash);
+            input_vec.push(right_hash);
+            if left_child_present && right_child_present {
+                both_children_present.push(true);
+            } else {
+                both_children_present.push(false);
+            }
+        }
+        // Process the input_vec using batch Poseidon hash
+        let output_vec = merkle_tree::field_based_mht::smt::big_lazy_merkle_tree::LazyBigMerkleTree::<F, T, P>::batch_poseidon_hash(input_vec);
+        // Place the computed hash in a cache_parallel
+        let mut index_output_vec = 0;
+        for coord in nodes_to_process_in_parallel[0].clone() {
+            self.state.cache_path.insert(coord, output_vec[index_output_vec]);
+            if both_children_present[index_output_vec] == true {
+                self.insert_to_cache(coord,output_vec[index_output_vec]);
+            } else {
+                self.remove_from_cache(coord);
+            }
+            index_output_vec += 1;
+        }
+
+        // Compute hashes in parallel - level > 1
```

Comment:
Can we merge this for with the previous "// Compute hashes in parallel - first level" code block ?

Reply:
Cleaned up a little bit, but merging the two code blocks doesn't really improve readability, so it's best to keep it as it is.